### PR TITLE
Introduce API to select and execute individual tests in inherited nested classes

### LINKED ISSUE #1913: Introduce API to select and execute individual tests in inherited nested classes
Such an API is necessary for tools to be able to run individual tests, when such tests are defined in a nested class of a (possibly abstract) super-class of the test class. See:

* https://youtrack.jetbrains.com/issue/IDEA-195373
* https://youtrack.jetbrains.com/issue/IDEA-215440

## Deliverables

I won't pretend to know how this should be done. Perhaps @akozlova has ideas.

### PR DESCRIPTION
**Pull request by juliette-derancourt**
Resolves #1913

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

File: junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/NestedClassSelector.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+ */
+@API(status = STABLE, since = "1.6")
+public class NestedClassSelector implements DiscoverySelector {
+
+	private List<ClassSelector> enclosingClassSelectors;
+	private ClassSelector nestedClassSelector;
+
+	NestedClassSelector(List<String> enclosingClassNames, String nestedClassName) {
+		this.enclosingClassSelectors = enclosingClassNames.stream().map(ClassSelector::new).collect(
+			Collectors.toList());
```
Comment (marcphilipp) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/NestedClassSelector.java]: That's an interesting trick! 👍

File: junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/NestedClassSelector.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+		this.nestedClassSelector = new ClassSelector(nestedClassName);
+	}
+
+	NestedClassSelector(List<Class<?>> enclosingClasses, Class<?> nestedClass) {
+		this.enclosingClassSelectors = enclosingClasses.stream().map(ClassSelector::new).collect(Collectors.toList());
+		this.nestedClassSelector = new ClassSelector(nestedClass);
+	}
+
+	public List<Class<?>> getEnclosingClasses() {
```
Comment (marcphilipp) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/NestedClassSelector.java]: We should add Javadoc to all public methods.

File: junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/NestedClassSelector.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+ * {@link #getEnclosingClasses()} or {@link #getNestedClass()} are invoked.
+ *
+ * <p>In this context, Java {@link Class} means anything that can be referenced
+ * as a {@link Class} on the JVM &mdash; for example, classes from other JVM
+ * languages such Groovy, Scala, etc.
+ *
+ * @since 1.6
+ * @see DiscoverySelectors#selectNestedClass(List, Class)
+ * @see DiscoverySelectors#selectNestedClass(List, String)
+ * @see org.junit.platform.engine.support.descriptor.ClassSource
```
Comment (marcphilipp) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/NestedClassSelector.java]: Should we point to `ClassSelector` as well?

File: junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/NestedClassSelector.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+		NestedClassSelector that = (NestedClassSelector) o;
+		return enclosingClassSelectors.equals(that.enclosingClassSelectors)
+				&& nestedClassSelector.equals(that.nestedClassSelector);
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(enclosingClassSelectors, nestedClassSelector);
+	}
+
```
Comment (marcphilipp) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/NestedClassSelector.java]: Please add `toString()` methods to both new selectors (see the existing selectors for examples).

File: platform-tests/src/test/java/org/junit/platform/engine/discovery/DiscoverySelectorsTests.java [hunk 5]
```diff
@@ -55,9 +60,9 @@
 
 	@Test
 	void selectUriByName() {
-		assertThrows(PreconditionViolationException.class, () -> selectUri((String) null));
-		assertThrows(PreconditionViolationException.class, () -> selectUri("   "));
-		assertThrows(PreconditionViolationException.class, () -> selectUri("foo:"));
+		assertViolatesPrecondition(() -> selectUri((String) null));
+		assertViolatesPrecondition(() -> selectUri("   "));
+		assertViolatesPrecondition(() -> selectUri("foo:"));
```
Comment (marcphilipp) [inline platform-tests/src/test/java/org/junit/platform/engine/discovery/DiscoverySelectorsTests.java]: 👍

File: junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/NestedMethodSelector.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+ * as a {@link Method} on the JVM &mdash; for example, methods from Java classes
+ * or methods from other JVM languages such Groovy, Scala, etc.
+ *
+ * @since 1.6
+ * @see DiscoverySelectors#selectNestedMethod(List, String, String)
+ * @see DiscoverySelectors#selectNestedMethod(List, String, String, String)
+ * @see DiscoverySelectors#selectNestedMethod(List, Class, String)
+ * @see DiscoverySelectors#selectNestedMethod(List, Class, String, String)
+ * @see DiscoverySelectors#selectNestedMethod(List, Class, Method)
+ * @see org.junit.platform.engine.support.descriptor.MethodSource
```
Comment (marcphilipp) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/NestedMethodSelector.java]: ... and to `MethodSelector` here.

File: junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/NestedClassSelector.java (hunk 3 above)
Comment (juliette-derancourt) [reply junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/NestedClassSelector.java]: You're right! 👍

File: junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/NestedMethodSelector.java [hunk 7]
```diff
@@ -154,4 +155,13 @@ public int hashCode() {
 		return Objects.hash(nestedClassSelector, methodSelector);
 	}
 
+    @Override
+    public String toString() {
+        return new ToStringBuilder(this) //
+                .append("enclosingClassNames", getEnclosingClassNames()) //
+                .append("nestedClassName", getNestedClassName()) //
+                .append("methodName", getMethodName()) //
+                .toString();
```
Comment (marcphilipp) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/NestedMethodSelector.java]: Please add `methodParameterTypes`.

Comment (juliette-derancourt) [reply junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/NestedMethodSelector.java]: Should I also add `getMethodParameterTypes()` then? I was wondering about that

Comment (marcphilipp) [reply junit-platform-engine/src/main/java/org/junit/platform/engine/discovery/NestedMethodSelector.java]: I guess so. 👍

**Comment by sormuras** [comment]
👍

**Comment by marcphilipp** [comment]
🎉

**Comment by juliette-derancourt** [comment]
😄

**Pull request merged**